### DISCUSSION TITLE
Is there a way to get the rule name from within a rule?

### DISCUSSION THREAD

**Discussion by djg2002** [hyperjumptech/grule-rule-engine#422] [General]
I would like to pass rule names to my functions which I can do with a string but its fragile.

Is there an object I can reference from within a rule that contains the current rule/context?  This would be really useful for establishing where behaviour is coming from.

Thanks

**Comment by tlaraMQ** [comment]
I agree and I think I want to do the same thing. Is it possible to retrieve the name of the activated rule?

**Comment by ivoshm** [comment]
As I know, it's different in 'when' and 'then' section of the rule:
- in 'then' section you can implement method EvaluateRuleEntry() in GruleEngineListener interface and store entry.RuleName in any accessible variable
- in 'when' section you face too aggressive caching - method getting RuleName is called only once for a fact and so it returns name of the first evaluated rule until the end of the cycle

**Comment by tlaraMQ** [comment]
Thank you for your answer. In my case is the 'then' section. I think  @djg2002  's is the 'when' section.

Could you elaborate on the GruleEngineListener interface and the EvaluateRuleEntry() method? I do not see any documentation regarding it.

**Reply by ivoshm** [reply]
I use it like this:

```golang

type RuleMachine struct {
        ...
	state        *RuleState
        ...
}

type RuleState struct {
        ...
	RuleName   string
        ...
}

func (r *RuleMachine) EvaluateRuleEntry(cycle uint64, entry *ast.RuleEntry, candidate bool) {
	if candidate {
		r.state.RuleName = entry.RuleName
	}
}
```

If you publish the instance of RuleState in the dataContext (something like `dataContext.Add("S", r.state)`) then you can get property S.RuleName in the rule.

Basic documentation of GruleEngineListener is available in source code:
https://github.com/hyperjumptech/grule-rule-engine/blob/master/engine/GruleEngineListener.go

I apologise for the brevity, but I hope this is understandable.

**Reply by ivoshm** [reply]
One more idea ... if you need RuleName only in "then" section it's enough to implement method ExecuteRuleEntry() like

```golang
func (r *RuleMachine) ExecuteRuleEntry(cycle uint64, entry *ast.RuleEntry) {
	r.state.RuleName = entry.RuleName
}
```
and don't "mess" with EvaluateRuleEntry()

**Reply by ivoshm** [reply]
But this whole approach is a hack - surely it could be done more elegantly and reliably inside GRL engine.

**Reply by tlaraMQ** [reply]
Thank you for your insights, I'll check if I can implement the approach
